header/p2p/peer_tracker.go

```diff
@@ -0,0 +1,101 @@
+package p2p
+
+import (
+	"context"
+	"sync"
+
+	"github.com/libp2p/go-libp2p-core/event"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+type peerTracker struct {
+	sync.RWMutex
+	connectedPeers map[peer.ID]*peerStat
+	// we should store peer to cache when it will be disconnected,
+	// so we can guarantee that peerQueue will return only active peer
+	disconnectedPeers map[peer.ID]*peerStat
+
+	host host.Host
+}
+
+func newPeerTracker(h host.Host) *peerTracker {
+	return &peerTracker{
+		disconnectedPeers: make(map[peer.ID]*peerStat),
+		connectedPeers:    make(map[peer.ID]*peerStat),
+		host:              h,
+	}
+}
+
+func (p *peerTracker) track(ctx context.Context) {
+	// store peers that have been already connected
+	for _, peer := range p.host.Peerstore().Peers() {
+		p.connected(peer)
+	}
+
+	subs, err := p.host.EventBus().Subscribe(&event.EvtPeerConnectednessChanged{})
+	if err != nil {
+		log.Errorw("subscribing to EvtPeerConnectednessChanged", "err", err)
+		return
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			err = subs.Close()
+			if err != nil {
+				log.Errorw("closing subscription", "err", err)
+			}
+			return
+		case subscription := <-subs.Out():
+			ev := subscription.(event.EvtPeerConnectednessChanged)
+			switch ev.Connectedness {
+			case network.Connected:
+				p.connected(ev.Peer)
+			case network.NotConnected:
+				p.disconnected(ev.Peer)
+			}
+		}
+	}
+}
+
+func (p *peerTracker) connected(pID peer.ID) {
+	if p.host.ID() == pID {
+		return
+	}
+	for _, c := range p.host.Network().ConnsToPeer(pID) {
+		if c.Stat().Transient {
+			return
+		}
+	}
+	p.Lock()
+	defer p.Unlock()
+	stats, ok := p.disconnectedPeers[pID]
+	if !ok {
+		stats = &peerStat{peerID: pID}
```

Comment:
yes, that makes sense, because we can have multiple connection events for the same peer. I would suggest:

```
if _, ok := p.connectedPeers; ok {
   return
}
```

Reply:
it can happen only in case when libp2p will emit EvtPeerConnectednessChanged.Connected multiple times with the same peerID. But iirc libp2p does no do it. cc @Wondertan

Reply:
i would document the assumption that this owuldn't be the case in the code @vgonkivs

Reply:
All the possible combinations of this event are documented here. Pls double check the assumption https://github.com/libp2p/go-libp2p/blob/master/core/event/network.go#L50